Components: Show tooltip only after mouseover delay

aduth opened the pull request:
Related: #2366

This pull request seeks to introduce a short delay when mousing over a Tooltip'd component before showing the tooltip text. Currently it is configured as a one second delay. The delay does not occur when the Tooltip'd component is focused by keyboard navigation.

__Testing instructions:__

Verify that tooltips show after a brief delay when moused over, and that they show immediately when tabbed to receive focus.

Examples: Content inserter button, block trash and cog, Undo/Redo, Close post settings sidebar

aduth committed:
Components: Show tooltip only after mouseover delay

timmyc reviewed (approved):
This is working well for me, and great work with the tests too.  I almost think 1 second is too long, maybe 500-750 might work well.

timmyc reviewed (commented):
Actually just found a slight glitch, note the tooltip turning off when i move the mouse, even though the B is still hovered:

![glitch](https://user-images.githubusercontent.com/22080/29291810-a0e9f6b6-80f9-11e7-92ef-071203eee59c.gif)

not sure if this is a regression though

aduth commented:
Hmm, could be related to event bubbling by `onmouseover` (i.e. the distinction between it and onmouseenter), causing it to become unopened immediately while mousing over elements within (the SVG icon).

timmyc commented:
Yeah not sure - oddly enough the button is still highlighted - in a mouseover state.  It _seems_ like the time the tooltip is hidden is about 1 second too.

aduth commented:
This appears to be made more difficult by a React bug: https://github.com/facebook/react/issues/4251

Specifically, it'd be better for us to use `mouseenter` and `mouseleave` events here, but they do not fire consistently for disabled elements. Default browser behavior is not to fire them at all; in React, at least `mouseenter` fires.

I considered a handful of alternatives:

- Supposedly with [`pointer-events` styles](https://github.com/facebook/react/issues/4251#issuecomment-267004045), one can force `mouseleave` events to fire. I'm not able to recreate this.
- Wrap the tooltip'd component with a non-disabled wrapper that attaches the event listeners. This could have an impact on style.
- Avoid using `disabled` attribute and instead "fake" it with `aria-disabled` attributes and equivalent styles. This appears to be what GMail uses for their fake buttons.
- Don't have tooltips for disabled buttons

timmyc commented:
Ugh kind of a mess then. I seem to recall similar challenges in Calypso with tooltips, but I'd have to dig a bit since the brain is not responding to the query at the moment.  If this is not a regression seems fine to merge this in as-is, or what are your thoughts?

aduth commented:
I think long-term, it's probably in our interest to refactor the `Button` component to render as a `div` emulating button characteristics. This helps avoid restrictions on children, and behaviors like the one here where mouse events do not fire on the disabled buttons.

I'd looked into this once before, and it's not trivial, since one must recreate all of the behaviors of a button manually (i.e. managing focus, triggering `onClick` only when non-disabled, etc). Certainly out of scope of this pull request.

I'm not really sure what to do with this pull request quite yet, since I don't find the behavior in this branch to be very good, though if the delay is considered important enough, still perhaps an improvement over what exists in `master`.

cc @afercia

afercia commented:
> This helps avoid restrictions on children

Not sure what children a button should have other than the button text 🙂 Ideally, buttons should be just buttons and not contain other HTML. Technically, it is possible to use a div to emulate a button but that would mean re-building all the native button interaction, not sure why we should do that in the first place. In core, we're trying to progressively [replace non-semantic buttons with real buttons](https://core.trac.wordpress.org/query?keywords=~semantic-buttons) because native UI controls are always preferable. We should always keep in mind that while emulated buttons might work visually and to solve a specific issue, they might not work in other cases, for example for other software that expect native controls to interact with.
I'd rather keep native buttons and not show tooltips for disabled buttons.

aduth commented:
That's ultimately what it comes down to: Either we don't show tooltips for disabled buttons, or we render it as an emulated button using another tag name. If we're content with the former, I can easily implement that here.

Generally though there's a distinction to be made between `Button` (a React component) and `button` (an HTML tag), where in a React component we're not bound to DOM specifics, and the underlying behavior is an implementation detail of little concern to the end-developer. We still have semantics; but they're the semantics of our component domain, not those dictated by the DOM specifications.

aduth committed:
Components: Use mouseenter for tooltip toggle

Mouse events behave unreliably in React for disabled elements, firing on mouseenter but not mouseleave. Further, the default behavior for disabled elements in some browsers is to ignore mouse events. Don't bother trying to to handle them.

See: https://github.com/facebook/react/issues/4251

aduth commented:
Updated in 4209e32 to use `mouseenter`, meaning in some (not all, see Firefox) browsers the tooltips will not be shown on disabled elements. I'd also explored using [pointer events](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events) `pointerenter` which had promising results in Chrome, but (a) [is not supported in React natively](https://github.com/facebook/react/issues/499) and (b) [has poor browser support generally](http://caniuse.com/#feat=pointer). I had been willing to accept it as a progressive enhancement, but really it only benefitted Chrome for showing tooltips on disabled elements, and did not seem worth supporting the inconsistency.

Reference implementation if something to be explored again in the future:

https://gist.github.com/aduth/bbb4405927ecc855a37fdbfc05e96ed6
